http.c

```diff
@@ -465,8 +513,8 @@ evhttp_make_header_request(struct evhttp_connection *evcon,
 	    method, req->uri, req->major, req->minor);
 
 	/* Add the content length on a post or put request if missing */
-	if ((req->type == EVHTTP_REQ_POST || req->type == EVHTTP_REQ_PUT) &&
-	    evhttp_find_header(req->output_headers, "Content-Length") == NULL){
+	if ((flags & EVHTTP_METHOD_HAS_BODY) &&
+	    evhttp_find_header(req->output_headers, "Content-Length") == NULL) {
```

Comment:
hum this patch may break something as more methods are defined to "have a body" because they could (but sometimes shoudn't)

Reply:
maybe we could have flags EVHTTP_METHOD_MAY_HAS_BODY and EVHTTP_METHOD_MUST_HAS_BODY instead of just EVHTTP_METHOD_HAS_BODY

Reply:
>hum this patch may break something as more methods are defined to "have a body" because they could (but sometimes shoudn't)

Indeed

>maybe we could have flags EVHTTP_METHOD_MAY_HAS_BODY and EVHTTP_METHOD_MUST_HAS_BODY instead of just EVHTTP_METHOD_HAS_BODY

 I didn't find anything in [rfc2616](https://tools.ietf.org/html/rfc2616) that POST "MUST" have body, hence I don't like this naming.

And this logic looks indeed strange, I would replace it with `if (evhttp_method_may_have_body(req->type) && !evhttp_find_header(req->output_headers, "Content-Length") && evbuffer_get_length(req->output_buffer)) ...` (but requires some testing)

Reply:
I rewrote the test in a way it won't break anything (forcing Content-Length with PUT and POST)

Reply:
>forcing Content-Length with PUT and POST

I would vote to avoid forcing Content-Length for PUT/POST, since it is valid to send POST without Content-Length

Reply:
well this is a feature dating back to a3bb4a035
https://github.com/libevent/libevent/commit/a3bb4a035fe27dcdf52158d115c4caea153dabf5#diff-f6b6aab933915430c8e4f4a5dfa99eb2R201
I don't think we should change it. Or then in a specific commit

Reply:
Agree